### ISSUE TITLE
Poor performance with many markers

### ISSUE THREAD

**Issue by zdila** [Leaflet/Leaflet#6314]
- [x] I'm reporting a bug, not asking for help; support questions like "How can I do X with Leaflet?" will be closed (use [Stack Overflow](http://stackoverflow.com/) or [gis.stackexchange.com](http://gis.stackexchange.com/) for questions)
- [x] I've looked at the [documentation](http://leafletjs.com/reference.html) to make sure the behaviour is documented and expected
- [x] I'm sure this is a Leaflet code issue, not an issue with my own code nor with the framework I'm using (Cordova, Ionic, Angular, React…)
- [x] I've searched through the issues to make sure it's not yet reported

## How to reproduce

- Leaflet version I'm using: 1.3.4
- Browser (with version) I'm using: Chromium 68
- OS/Platform (with version) I'm using: Debian Buster
- Place many markers to the map (eg. 1000)
- Move or zoom the map

## What behaviour I'm expecting and which behaviour I'm seeing

Start of the movement is delayed and eats 100% of the CPU core. The same for the zoom. Delay si caused by `getBoundingClientRect`:

![image](https://user-images.githubusercontent.com/636095/44630020-d7a58d00-a957-11e8-97b1-839a71fe0955.png)

## Minimal example reproducing the issue

- [x] this example is as simple as possible
- [x] this example does not rely on any third party code

https://codepen.io/zdila/pen/YOwoEd

**Comment by ghybs** [comment]
Hi @zdila,

Unfortunately that is a known limitation of Leaflet.
See https://stackoverflow.com/questions/43015854/large-dataset-of-markers-or-dots-in-leaflet/43019740#43019740

Leaflet uses regular DOM Elements for standard Markers.
This simplicity has a drawback: browsers struggle handling thousands of them.

For the common usage (a few Markers), the simplicity is advantageous.

For more complex use case, the workaround is to switch to canvas.
There should be some plugins that provide other canvas-based markers, should you need something fancier than Circle Markers.

**Comment by manubb** [comment]
If you need to draw many markers, you can try Leaflet.PixiOverlay [plugin](https://github.com/manubb/Leaflet.PixiOverlay).
(i'm the plugin author by the way)

**Comment by zdila** [comment]
@ghybs thanks for the explanation. I just wonder if calling `getBoundingClientRect` is really necessary and what is it used for in Leaflet.

@manubb thaks for the info. We already use canvas for displaying picture positions on the map: https://www.freemap.sk/?map=9/48.821333/20.126953&layers=TI

**Comment by php4fan** [comment]
> For the common usage (a few Markers)

That's the key wrong assumption.

**Comment by php4fan** [comment]
It's time for this to be reopened (actually it shouldn't have been closed in the first place).

The idea that the normal use case for a map is to have less than a few dozens points was already wrong ten years ago, but it's absolutely laughable today.

Thousands of points _is_ the common use case.

> Unfortunately that is a known limitation of Leaflet.

Then it has to be addressed because it's an acceptable limitation and solutions exist (Google Maps doesn't have this issue).

> Leaflet uses regular DOM Elements for standard Markers.
> This simplicity has a drawback: browsers struggle handling thousands of them.

Then this simplicity is not viable. Thousands of markers is a bare minimum requirement.

> For the common usage (a few Markers), the simplicity is advantageous.

A few markers (meaning like a dozen) is not common usage. If that simplicity is advantageous in extremely rare, minimal use cases such as maps with only a handful of markers, then offer it as an option, but not the default, let alone the only option.

> For more complex use case, the workaround is to switch to canvas.
> There should be some plugins that provide other canvas-based markers

One should not have to look for plugins to have a basic simple map work out of the box. Again, thousands of markers is a typical use case.
If that needs canvas, then canvas should be the default (giving the developer the possibility to override it and force DOM), and/or you should switch transparently and seamlessly between canvas and DOM depending on the number of markers (or whatever criteria make it necessary), again obviously by default and leaving the developer the choice to force either one or the other). That is assuming that there is truly no way to get DOM to work smoothly.

The bottom line is, a map with thousands of markers (which is a common, basic use case) should work out of the box with no plugins, no special configuration, no workaround. Whether that means you need to use canvas by default or whether it can be achieved with other kinds of optimizations is up to you.

**Comment by dresende** [comment]
@php4fan is there any progress on this? can you contribute with your ideas or changes?

**Comment by php4fan** [comment]
> @php4fan is there any progress on this?

LOL I have no idea, I'm not a contributor here. But it doesn't seem likely, since they don't even seem to understand this as an issue. 🤷

> can you contribute with your ideas

I don't think any novel "idea" is needed here. They just need to understand that hundreds or thousands of markers _is_ the common use case for a modern map, and a few dozens of markers or less is the edge case; not the other way around.

Solutions to have many markers work smoothly have already been invented. They say canvas-based markers is one of them: if that's true, then they just need to make that the default, and more importantly, make it 100% transparent, i.e. you just enable canvas by setting a boolean switch to true with no additional complications (once you have that, it's less important whether or not that's the default, although it should be).

**Comment by dresende** [comment]
I was sarcastic, I was not expecting a contribution other than bashing. It doesn't seem like a very constructive comment to just expect people on their free time to solve your problems.

I too am affected by this issue but I would rather help or look for an alternative. You comment was unnecessary.

**Comment by php4fan** [comment]
>

**Comment by dresende** [comment]
https://gitlab.com/IvanSanchez/Leaflet.GLMarkers
